In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/datathon/Total_Electricity_Production.csv
/kaggle/input/datathon/south_africa_temperature.png
/kaggle/input/datathon/south_africa_load_shedding_history.csv
/kaggle/input/datathon/world_indicators.csv


What is the relationship between Electricity production and Load_shedding in South Africa?

In [2]:
import numpy as np
import datetime
import plotly.graph_objs as go 
import plotly.io as pio 
import plotly.express as px 

# Importing Plotly 
#import plotly.offline as py 
#py.init_notebook_mode(connected=True) 
  
# Initializing Plotly 
pio.renderers.default = 'colab'

In [3]:
df = pd.read_csv("/kaggle/input/datathon/world_indicators.csv")
df2 = pd.read_csv("/kaggle/input/datathon/Total_Electricity_Production.csv")
df3 = pd.read_csv("/kaggle/input/datathon/south_africa_load_shedding_history.csv")

**Lets check the first five rows of df**

In [4]:
df.head()

Country Name Country Code  \
0  South Africa          ZAF   
1  South Africa          ZAF   
2  South Africa          ZAF   
3  South Africa          ZAF   
4  South Africa          ZAF   

                                         Series Name        Series Code  \
0  Population density (people per sq. km of land ...        EN.POP.DNST   
1                                  Population, total        SP.POP.TOTL   
2                                  GDP (current US$)     NY.GDP.MKTP.CD   
3                              GDP growth (annual %)  NY.GDP.MKTP.KD.ZG   
4                 GDP per capita (constant 2010 US$)     NY.GDP.PCAP.KD   

      2000 [YR2000]     2001 [YR2001]     2002 [YR2002]     2003 [YR2003]  \
0  37.0687319160161  37.5662762037442  38.0440964808877  38.5125555399847   
1          44967708          45571274          46150913          46719196   
2  136361298082.061  121514658737.165  115482368343.658  175256916996.047   
3  4.20000347551795  2.69999456713741  3.70038235175596   2.9490791374903   
4   5937.6260860986  6017.17784586121  6161.46631380485  6266.01573980923   

      2004 [YR2004]     2005 [YR2005]  ...     2011 [YR2011]  \
0  38.9844199523531  39.4699494678878  ...  42.8691787089169   
1          47291610          47880601  ...          52004172   
2  228590027400.653  257772710832.953  ...  416418874936.304   
3  4.55455274472709   5.2770563116366  ...   3.2841681423114   
4  6472.10703408701   6729.8274660259  ...  7454.72016478301   

      2012 [YR2012]     2013 [YR2013]     2014 [YR2014]     2015 [YR2015]  \
0  43.5532441945775  44.2582462966474  44.9645046946228  45.6572612089787   
1          52834005          53689236          54545991          55386367   
2  396332702639.496  366829390478.954  350904575292.317  317620522794.827   
3  2.21335480848441   2.4852005003106  1.84699160365717  1.19373280124428   
4  7500.04106663005  7563.99277707639  7582.69692889496  7556.78857882235   

       2016 [YR2016]     2017 [YR2017]      2018 [YR2018]      2019 [YR2019]  \
0   46.3309845106299   46.987817062213   47.6301197767684                 ..   
1           56203654          57000451           57779622           58558270   
2   296357282715.109  349554116683.818   368288939768.322   351431649241.439   
3  0.399087929556359  1.41451262585058  0.787055570495099  0.152583316664234   
4   7476.62101155809  7476.38632385801   7433.61539826952   7345.96241792901   

  2020 [YR2020]  
0            ..  
1            ..  
2            ..  
3            ..  
4            ..  

[5 rows x 25 columns]

we will be focusing on the df from 2015 to recent data available so lets drop other years

In [5]:
df00 = df[['Country Name', 'Country Code','Series Code','2015 [YR2015]','2016 [YR2016]','2017 [YR2017]','2018 [YR2018]','2019 [YR2019]']]

In [6]:
df00

Country Name Country Code  \
0                                        South Africa          ZAF   
1                                        South Africa          ZAF   
2                                        South Africa          ZAF   
3                                        South Africa          ZAF   
4                                        South Africa          ZAF   
..                                                ...          ...   
228                                               NaN          NaN   
229                                               NaN          NaN   
230                                               NaN          NaN   
231  Data from database: World Development Indicators          NaN   
232                          Last Updated: 09/16/2020          NaN   

           Series Code     2015 [YR2015]      2016 [YR2016]     2017 [YR2017]  \
0          EN.POP.DNST  45.6572612089787   46.3309845106299   46.987817062213   
1          SP.POP.TOTL          55386367           56203654          57000451   
2       NY.GDP.MKTP.CD  317620522794.827   296357282715.109  349554116683.818   
3    NY.GDP.MKTP.KD.ZG  1.19373280124428  0.399087929556359  1.41451262585058   
4       NY.GDP.PCAP.KD  7556.78857882235   7476.62101155809  7476.38632385801   
..                 ...               ...                ...               ...   
228                NaN               NaN                NaN               NaN   
229                NaN               NaN                NaN               NaN   
230                NaN               NaN                NaN               NaN   
231                NaN               NaN                NaN               NaN   
232                NaN               NaN                NaN               NaN   

         2018 [YR2018]      2019 [YR2019]  
0     47.6301197767684                 ..  
1             57779622           58558270  
2     368288939768.322   351431649241.439  
3    0.787055570495099  0.152583316664234  
4     7433.61539826952   7345.96241792901  
..                 ...                ...  
228                NaN                NaN  
229                NaN                NaN  
230                NaN                NaN  
231                NaN                NaN  
232                NaN                NaN  

[233 rows x 8 columns]

Lets see which countries are covered

In [7]:
df['Country Name'].unique()

array(['South Africa', 'Nigeria', 'Sub-Saharan Africa', 'World',
       'Egypt, Arab Rep.', 'Algeria', 'Morocco', 'Libya', 'Botswana',
       'Namibia', 'Gabon', 'Equatorial Guinea', nan,
       'Data from database: World Development Indicators',
       'Last Updated: 09/16/2020'], dtype=object)

Lets further filter the data and leave south africa's data then drop all other countries

In [8]:
df0 = df00.loc[df00['Country Name'] == 'South Africa']

In [9]:
df0.T

0             1                 2   \
Country Name       South Africa  South Africa      South Africa   
Country Code                ZAF           ZAF               ZAF   
Series Code         EN.POP.DNST   SP.POP.TOTL    NY.GDP.MKTP.CD   
2015 [YR2015]  45.6572612089787      55386367  317620522794.827   
2016 [YR2016]  46.3309845106299      56203654  296357282715.109   
2017 [YR2017]   46.987817062213      57000451  349554116683.818   
2018 [YR2018]  47.6301197767684      57779622  368288939768.322   
2019 [YR2019]                ..      58558270  351431649241.439   

                              3                 4                 5   \
Country Name        South Africa      South Africa      South Africa   
Country Code                 ZAF               ZAF               ZAF   
Series Code    NY.GDP.MKTP.KD.ZG    NY.GDP.PCAP.KD    EG.ELC.ACCS.ZS   
2015 [YR2015]   1.19373280124428  7556.78857882235              85.5   
2016 [YR2016]  0.399087929556359  7476.62101155809              84.2   
2017 [YR2017]   1.41451262585058  7476.38632385801              84.4   
2018 [YR2018]  0.787055570495099  7433.61539826952  91.2298736572266   
2019 [YR2019]  0.152583316664234  7345.96241792901                ..   

                              6                  7               8   \
Country Name        South Africa       South Africa    South Africa   
Country Code                 ZAF                ZAF             ZAF   
Series Code    EG.ELC.ACCS.RU.ZS  EG.ELC.ACCS.UR.ZS  EN.CO2.ETOT.ZS   
2015 [YR2015]   75.3635314242526   90.9994735717773              ..   
2016 [YR2016]   70.7157223672873   91.3525009155273              ..   
2017 [YR2017]   70.3145301045502   91.7047653198242              ..   
2018 [YR2018]   89.5988191888486   92.0568923950195              ..   
2019 [YR2019]                 ..                 ..              ..   

                             9                  10              11  \
Country Name       South Africa       South Africa    South Africa   
Country Code                ZAF                ZAF             ZAF   
Series Code      EG.ELC.COAL.ZS     EG.ELC.HYRO.ZS  EG.ELC.NGAS.ZS   
2015 [YR2015]  92.7112379223137  0.324638479994812               0   
2016 [YR2016]                ..                 ..              ..   
2017 [YR2017]                ..                 ..              ..   
2018 [YR2018]                ..                 ..              ..   
2019 [YR2019]                ..                 ..              ..   

                           12                 13                14  \
Country Name     South Africa       South Africa      South Africa   
Country Code              ZAF                ZAF               ZAF   
Series Code    EG.ELC.NUCL.ZS     EG.ELC.PETR.ZS    EG.ELC.FOSL.ZS   
2015 [YR2015]              ..  0.074168341871474  92.7854062641852   
2016 [YR2016]              ..                 ..                ..   
2017 [YR2017]              ..                 ..                ..   
2018 [YR2018]              ..                 ..                ..   
2019 [YR2019]              ..                 ..                ..   

                             15              16                17  \
Country Name       South Africa    South Africa      South Africa   
Country Code                ZAF             ZAF               ZAF   
Series Code      EG.ELC.RNWX.ZS  EG.ELC.RNWX.KH    EG.ELC.RNEW.ZS   
2015 [YR2015]  1.93040334608651      4763000000  2.25504182608132   
2016 [YR2016]                ..              ..                ..   
2017 [YR2017]                ..              ..                ..   
2018 [YR2018]                ..              ..                ..   
2019 [YR2019]                ..              ..                ..   

                         18  
Country Name   South Africa  
Country Code            ZAF  
Series Code     IC.ELC.TIME  
2015 [YR2015]            84  
2016 [YR2016]           109  
2017 [YR2017]           109  
2018 [YR2018]           109  
2019 [YR2019

After we have filtered the data lets check if there are any null values

In [10]:
df0.isnull().sum()

Country Name     0
Country Code     0
Series Code      0
2015 [YR2015]    0
2016 [YR2016]    0
2017 [YR2017]    0
2018 [YR2018]    0
2019 [YR2019]    0
dtype: int64

There are no empty cells we can move to the 2nd data which is called df2 that represents Electricity Production

**Lets check the first five rows of df2**

In [11]:
df2.head()

DATE  PRENEL01ZAQ656N
0  1985-01-01     11072.666667
1  1985-04-01     11879.666667
2  1985-07-01     12485.333333
3  1985-10-01     11690.333333
4  1986-01-01     11374.000000

Lets see the number of rows and colums on df2

In [12]:
df2.shape

(135, 2)

From the fist five rows we find that the data is from the year 1985 so lets check when does the date start and end

In [13]:
df2['DATE'].unique()

array(['1985-01-01', '1985-04-01', '1985-07-01', '1985-10-01',
       '1986-01-01', '1986-04-01', '1986-07-01', '1986-10-01',
       '1987-01-01', '1987-04-01', '1987-07-01', '1987-10-01',
       '1988-01-01', '1988-04-01', '1988-07-01', '1988-10-01',
       '1989-01-01', '1989-04-01', '1989-07-01', '1989-10-01',
       '1990-01-01', '1990-04-01', '1990-07-01', '1990-10-01',
       '1991-01-01', '1991-04-01', '1991-07-01', '1991-10-01',
       '1992-01-01', '1992-04-01', '1992-07-01', '1992-10-01',
       '1993-01-01', '1993-04-01', '1993-07-01', '1993-10-01',
       '1994-01-01', '1994-04-01', '1994-07-01', '1994-10-01',
       '1995-01-01', '1995-04-01', '1995-07-01', '1995-10-01',
       '1996-01-01', '1996-04-01', '1996-07-01', '1996-10-01',
       '1997-01-01', '1997-04-01', '1997-07-01', '1997-10-01',
       '1998-01-01', '1998-04-01', '1998-07-01', '1998-10-01',
       '1999-01-01', '1999-04-01', '1999-07-01', '1999-10-01',
       '2000-01-01', '2000-04-01', '2000-07-01', '2000-

Now lets filter the data from 2015 to the recent date available

In [14]:
df_2 = df2[ df2['DATE'] >= '2015-01-01' ]

In [15]:
df_2

DATE  PRENEL01ZAQ656N
120  2015-01-01     20714.666667
121  2015-04-01     20882.000000
122  2015-07-01     21197.000000
123  2015-10-01     20589.000000
124  2016-01-01     20469.333333
125  2016-04-01     21064.000000
126  2016-07-01     21761.666667
127  2016-10-01     20954.000000
128  2017-01-01     20448.333333
129  2017-04-01     21558.666667
130  2017-07-01     21765.666667
131  2017-10-01     21262.000000
132  2018-01-01     20823.000000
133  2018-04-01     21483.666667
134  2018-07-01     21837.333333

lets check if there are any empty cells

In [16]:
df_2.isnull().sum()

DATE               0
PRENEL01ZAQ656N    0
dtype: int64

There are no empty cells therefore we can move to the final dataset

**For this data lets check the sample of 5 random rows**

In [17]:
df3.sample(5)

created_at  stage
213  7/22/2015 17:02:26      1
255  6/20/2015 17:08:55      1
188   8/8/2015 16:58:18      1
105  11/7/2019 22:00:29      2
23   8/14/2020 19:31:25      0

lets check when does the date starts and end

In [18]:
df3['created_at'].unique()

array(['9/7/2020 20:03:18', '9/7/2020 14:00:22', '9/7/2020 6:00:53',
       '9/6/2020 18:16:14', '9/6/2020 10:00:25', '9/6/2020 6:00:50',
       '9/5/2020 20:00:31', '9/5/2020 6:00:18', '9/4/2020 19:12:05',
       '9/4/2020 6:01:47', '9/3/2020 20:04:09', '9/3/2020 6:00:35',
       '9/2/2020 20:21:48', '9/2/2020 13:00:13', '9/2/2020 6:00:50',
       '9/1/2020 20:00:23', '9/1/2020 10:00:34', '8/20/2020 20:00:04',
       '8/20/2020 7:02:58', '8/19/2020 20:00:55', '8/19/2020 7:00:20',
       '8/18/2020 19:56:07', '8/18/2020 14:01:02', '8/14/2020 19:31:25',
       '8/14/2020 15:00:49', '8/14/2020 5:59:58', '8/13/2020 20:02:24',
       '8/13/2020 5:55:23', '7/16/2020 19:15:17', '7/16/2020 12:00:23',
       '7/16/2020 7:00:32', '7/15/2020 19:50:37', '7/15/2020 12:01:40',
       '7/15/2020 7:00:28', '7/14/2020 19:57:28', '7/14/2020 7:03:33',
       '7/13/2020 20:01:22', '7/13/2020 7:00:26', '7/12/2020 20:00:55',
       '7/12/2020 18:00:42', '7/12/2020 5:59:43', '7/11/2020 20:34:30',
       '7/

the data starts from 2015 now lets check if there are any empty cells

In [19]:
df3.isnull().sum()

created_at    0
stage         0
dtype: int64

***VISUALIZATION***

Lets create a line graph for south Africa load_shedding represented by df3

In [20]:
fig = px.line(df3, x='created_at', y='stage', title='Load-shedding')
fig.show()

Now lets see how a pie graph show us df_2

In [21]:
fig = px.pie(df_2, values='PRENEL01ZAQ656N', names='DATE', title='Electricity-Production')
fig.show()

**Now lets see what the statistic says about the dataset**

In [22]:
df_2.describe().T

count          mean         std           min           25%  \
PRENEL01ZAQ656N   15.0  21120.688889  477.363351  20448.333333  20768.833333   

                     50%           75%           max  
PRENEL01ZAQ656N  21064.0  21521.166667  21837.333333

The results shows that the maximum value of electricity produced is 21837.333333 which was produced in 2019 according to the bar chart. The minimun value is 20448.333333 which was produced in 2015. The average is 21120.688889 and the standard deviation is 477.363351 which simply tell us that each year is approximately 478 produced away from the average

**Now we move to the stats of df3**

In [23]:
df3.describe().T

count      mean       std  min  25%  50%  75%  max
stage  422.0  1.059242  1.067239  0.0  0.0  1.0  2.0  6.0

Loadshedding results shows that stage 6 was the maximum stage reached in 2019 then the minimum stage is 0 which is dominating in 2015. stage 1.059242 which is stage 1 is the average and the distance from the average of each stage is 1.067239

***CONCLUSION***

At this point we experienced stage 6 only once in 2019 whereby electricity produced was higher. Ranging from 2015 to 2018 there was no much impact therefore the difference is indeed ignorable. from 2019 to 2020 there was a rise in stages of loadshedding. well the relationship is not proportional